### ISSUE TITLE
Cygwin: Run-shell fails in scripts with bashism

### ISSUE THREAD

**Issue by Xaldew** [tmux/tmux#940]
I've run into a somewhat odd issue on Cygwin when I'm using the Tmux command run-shell. Given the script:

    #!/usr/bin/env bash
    
    read -r -d '' var << EOF
    This is line 1.
    This is line 2.
    Line 3.
    EOF
    
    echo "Running in shell: $SHELL"
    
    while read line;
    do
        echo $line
    done <<EOF
    $var
    EOF
    
    while read line;
    do
        echo $line
    done < <(echo "$var")

Running it directly works just fine (`bash ~/test.bash`), producing:  

    Running in shell: /bin/bash
    This is line 1.
    This is line 2.
    Line 3.
    This is line 1.
    This is line 2.
    Line 3.

But when I try to run this using `tmux run-shell test.bash` I get:

    Running in shell: /bin/bash
    This is line 1.
    This is line 2.
    Line 3.
    '/cygdrive/c/Users/gusta/test.bash' returned 2

Now, I'm aware that the second loop is a Bashism, so I've probably missed some kind of configuration to ensure that Tmux properly uses Bash for the run-shell command but so far I've not been able to find such an option. I've tried setting:

    set-option -g default-shell /bin/bash
    set-option -g default-command /bin/bash

in `.tmux.conf` but that doesn't seem to be working.

Current configuration:

    $ uname -sp && tmux -V && echo $TERM
    CYGWIN_NT-10.0 unknown
    tmux 2.3
    tmux-24bit

(I've also tried this on tmux-2.2 and tmux-master with the same results, with and without my personal configuration.)

**Comment by nicm** [comment]
run-shell will always use /bin/sh -c, but if you are running the script directly Cygwin should use the #! line. Does it work if you do this from the shell:
~~~~
/bin/sh -c  /my/script

**Comment by Xaldew** [comment]
Running with `LC_ALL=C /bin/sh -c ~/test.bash` does indeed give me the same behaviour, although a slightly different error message:

```
Running in shell: /bin/bash
This is line 1.
This is line 2.
Line 3.
/cygdrive/c/Users/xaldew/test.bash: line 23: syntax error near unexpected token `<'
/cygdrive/c/Users/xaldew/test.bash: line 23: `done < <(echo "$var")'
````

But I guess that's just from dropping `STDERR` and displaying a custom error message.

I assume this particular behaviour isn't configurable? Apparently `Cygwin` sets `/bin/sh` to `bash` and as far as I could read from the bash manual, when it is invoked as `sh` it automatically runs in a Posix compliant mode. Although, I guess I could use a different shell as `/bin/sh` to fix that particular behaviour.

Additionally, does this mean that all `tmux` plugins should really be written as Posix shell scripts in order to be portable?

**Comment by nicm** [comment]
No. If you run a script that is executable, it should be run with bash if
that is what is in the #! line, _even if you initially start it with a
different shell_.

So assuming you are doing:

/bin/sh -c /my/script

And not one of these:

/bin/sh /my/script
/bin/sh -c sh /my/script

It should use bash.

This sounds like a Cygwin bug.



On 29 May 2017 10:22 pm, "Gustaf Waldemarson" <notifications@github.com>
wrote:

Closed #940 <https://github.com/tmux/tmux/issues/940>.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/tmux/tmux/issues/940#event-1101436962>, or mute the
thread

**Comment by Xaldew** [comment]
I agree. I'll move this question on to the Cygwin mailing list. Thanks for the info though and keep up the good work!
